// @preact/preset-vite/dist/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.preact = void 0;
const vite_1 = __importDefault(require("@prefresh/vite"));
const devtools_js_1 = require("./devtools.js");
const utils_js_1 = require("./utils.js");
const prerender_js_1 = require("./prerender.js");
const core_1 = require("@babel/core");
// Taken from https://github.com/vitejs/vite/blob/main/packages/plugin-react/src/index.ts
function preactPlugin({ devtoolsInProd, devToolsEnabled, prefreshEnabled, reactAliasesEnabled, prerender, include, exclude, babel, jsxImportSource, } = {}) {
    var _a;
    const baseParserOptions = [
        "importMeta",
        "explicitResourceManagement",
        "topLevelAwait",
    ];
    let config;
    let babelOptions = {
        babelrc: false,
        configFile: false,
        ...babel,
    };
    babelOptions.plugins || (babelOptions.plugins = []);
    babelOptions.presets || (babelOptions.presets = []);
    babelOptions.overrides || (babelOptions.overrides = []);
    babelOptions.parserOpts || (babelOptions.parserOpts = {});
    (_a = babelOptions.parserOpts).plugins || (_a.plugins = []);
    const shouldTransform = (0, utils_js_1.createFilter)(include || [/\.[tj]sx?$/], exclude || [/node_modules/]);
    devToolsEnabled = devToolsEnabled !== null && devToolsEnabled !== void 0 ? devToolsEnabled : true;
    prefreshEnabled = prefreshEnabled !== null && prefreshEnabled !== void 0 ? prefreshEnabled : true;
    reactAliasesEnabled = reactAliasesEnabled !== null && reactAliasesEnabled !== void 0 ? reactAliasesEnabled : true;
    prerender = prerender !== null && prerender !== void 0 ? prerender : { enabled: false };
    const jsxPlugin = {
        name: "vite:preact-jsx",
        enforce: "pre",
        config() {
            return {
                build: {
                    rollupOptions: {
                        onwarn(warning, warn) {
                            // Silence Rollup's module-level directive warnings -- they're likely
                            // to all come from `node_modules` (RSCs) and won't be actionable.
                            if (warning.code === "MODULE_LEVEL_DIRECTIVE")
                                return;
                            warn(warning);
                        },
                    },
                },
                optimizeDeps: {
                    include: ["preact/jsx-runtime", "preact/jsx-dev-runtime"],
                },
            };
        },
        configResolved(resolvedConfig) {
            config = resolvedConfig;
        },
        async transform(code, url) {
            // Ignore query parameters, as in Vue SFC virtual modules.
            const { id } = (0, utils_js_1.parseId)(url);
            if (!shouldTransform(id))
                return;
            const parserPlugins = [
                ...baseParserOptions,
                "classProperties",
                "classPrivateProperties",
                "classPrivateMethods",
                !id.endsWith(".ts") && "jsx",
                /\.tsx?$/.test(id) && "typescript",
            ].filter(Boolean);
            const result = await (0, core_1.transformAsync)(code, {
                ...babelOptions,
                ast: true,
                root: config.root,
                filename: id,
                parserOpts: {
                    ...babelOptions.parserOpts,
                    sourceType: "module",
                    allowAwaitOutsideFunction: true,
                    plugins: parserPlugins,
                },
                generatorOpts: {
                    ...babelOptions.generatorOpts,
                    decoratorsBeforeExport: true,
                },
                plugins: [
                    ...babelOptions.plugins,
                    [
                        config.isProduction
                            ? "@babel/plugin-transform-react-jsx"
                            : "@babel/plugin-transform-react-jsx-development",
                        {
                            runtime: "automatic",
                            importSource: jsxImportSource !== null && jsxImportSource !== void 0 ? jsxImportSource : "preact",
                        },
                    ],
                    ...(devToolsEnabled && !config.isProduction ? ["babel-plugin-transform-hook-names"] : []),
                ],
                sourceMaps: true,
                inputSourceMap: false,
            });
            // NOTE: Since no config file is being loaded, this path wouldn't occur.
            if (!result)
                return;
            return {
                code: result.code || code,
                map: result.map,
            };
        },
    };
    return [
        ...(reactAliasesEnabled
            ? [
                {
                    name: "preact:config",
                    config() {
                        return {
                            resolve: {
                                alias: {
                                    "react-dom/test-utils": "preact/test-utils",
                                    "react-dom": "preact/compat",
                                    react: "preact/compat",
                                },
                            },
                        };
                    },
                },
            ]
            : []),
        jsxPlugin,
        ...(devToolsEnabled
            ? [
                (0, devtools_js_1.preactDevtoolsPlugin)({
                    injectInProd: devtoolsInProd,
                    shouldTransform,
                }),
            ]
            : []),
        ...(prefreshEnabled
            ? [(0, vite_1.default)({ include, exclude, parserPlugins: baseParserOptions })]
            : []),
        ...(prerender.enabled ? [(0, prerender_js_1.PrerenderPlugin)(prerender)] : []),
    ];
}
exports.preact = preactPlugin;
exports.default = preactPlugin;
